mpqp/qasm/mpqp_to_qasm.py

```diff
@@ -0,0 +1,218 @@
+    final_str = ""
+
+    while any(target != 0 for target in targets.values()):
+        final_str += "\n" + instruction_str + " "
+        if all(target != 0 for target in targets.values()):
+            final_str += "q;"
+            targets = {key: value - 1 for key, value in targets.items()}
+        else:
+            for key, target in targets.items():
+                if target != 0:
+                    final_str += f"q[{key}],"
+                    targets[key] -= 1
+            final_str = final_str[:-1] + ";"
+
+    return final_str
+
+
+def mpqp_to_qasm2(circuit: QCircuit, simplify: bool = False) -> str:
+    """
+    Converts an MPQP `QCircuit` object into a string in QASM 2.0 format.
+    It handles various quantum instructions like gates, measurements, and barriers and
+    can optionally simplify the circuit by merging consecutive single-qubit gates of the same type.
+
+    Args:
+        circuit: The MPQP quantum circuit to be converted.
+        simplify:  If set to True, the function will attempt to simplify the circuit
+        by merging consecutive single-qubit gates of the same type, if applicable.
+
+    Returns:
+        A string containing the QASM 2.0 representation of the provided quantum circuit.
+
+    Raises:
+        ValueError: If an unknown gate or instruction type is encountered during the conversion process.
+
+    Example:
+        >>> circuit = QCircuit([H(0), CNOT(0, 1), BasisMeasure()])
+        >>> qasm_code = mpqp_to_qasm2(circuit, simplify=True)
+        >>> print(qasm_code)
+        OPENQASM 2.0;
+        include "qelib1.inc";
+        qreg q[2];
+        creg c[2];
+        h q[0];
+        cx q[0],q[1];
+        measure q[0] -> c[0];
+        measure q[1] -> c[1];
+    """
+    circuit_copy = circuit.hard_copy()
+    if circuit_copy.noises:
+        logging.warning(
+            "Instructions such as noise are not supported by QASM2 hence have been ignored."
+        )
+
+    qasm_str = (
+        "OPENQASM 2.0;"
+        + "\ninclude \"qelib1.inc\";"
+        + f"\nqreg q[{circuit_copy.nb_qubits}];"
+    )
+
+    if circuit_copy.nb_cbits != None:
+        qasm_str += f"\ncreg c[{circuit_copy.nb_cbits}];"
+
+    previous = None
+    targets = {i: 0 for i in range(circuit_copy.nb_qubits)}
+
+    for instruction in circuit_copy.instructions:
+        if simplify:
+            if isinstance(instruction, SingleQubitGate):
+                if previous is None or not isinstance(instruction, type(previous)):
+                    if previous:
+                        qasm_str += _simplify_instruction(previous, targets)
+                        targets = {i: 0 for i in range(circuit_copy.nb_qubits)}
+                    previous = instruction
+                else:
+                    if (
+                        isinstance(instruction, ParametrizedGate)
+                        and instruction.parameters
+                        != previous.parameters  # pyright: ignore[reportAttributeAccessIssue]
```

Comment:
I would do the export in two steps, first an export using instructions'
`to_other_language` methods, then a simplification step, with simple rewriting step